topostats/tracing/disordered_tracing.py

```diff
@@ -0,0 +1,793 @@
+
+        This method ensures gaussian smoothing isn't too aggressive and covers / creates gaps in the mask.
+
+        Parameters
+        ----------
+        grain : npt.NDArray
+            Numpy array of the grain mask.
+        dilation_iterations : int
+            Number of times to dilate the grain to smooth it. Default is 2.
+        gaussian_sigma : float | None
+            Gaussian sigma value to smooth the grains after an Otsu threshold. If None, defaults to 2.
+        holearea_min_max : tuple[float | int | None]
+            Tuple of minimum and maximum hole area (in nanometers) to replace from the original mask into the smoothed
+            mask.
+
+        Returns
+        -------
+        npt.NDArray
+            Numpy array of smmoothed image.
+        """
+        dilation = ndimage.binary_dilation(grain, iterations=dilation_iterations).astype(np.int32)
+        gauss = filters.gaussian(grain, sigma=gaussian_sigma)
+        gauss = np.where(gauss > filters.threshold_otsu(gauss) * 1.3, 1, 0)
+        gauss = gauss.astype(np.int32)
+        # Add hole to the smooth mask conditional on smallest pixel difference for dilation or the Gaussian smoothing.
+        if dilation.sum() > gauss.sum():
+            LOGGER.debug(f"[{self.filename}] : smoothing done by gaussian {gaussian_sigma}")
+            return self.re_add_holes(grain, gauss, holearea_min_max)
+        LOGGER.debug(f"[{self.filename}] : smoothing done by dilation {dilation_iterations}")
+        return self.re_add_holes(grain, dilation, holearea_min_max)
+
+
+def trace_image_disordered(  # pylint: disable=too-many-arguments,too-many-locals
+    image: npt.NDArray,
+    grains_mask: npt.NDArray,
+    filename: str,
+    pixel_to_nm_scaling: float,
+    min_skeleton_size: int,
+    mask_smoothing_params: dict,
+    skeletonisation_params: dict,
+    pruning_params: dict,
+    pad_width: int = 1,
+) -> dict:
+    """
+    Processor function for tracing image.
+
+    Parameters
+    ----------
+    image : npt.NDArray
+        Full image as Numpy Array.
+    grains_mask : npt.NDArray
+        Full image as Grains that are labelled.
+    filename : str
+        File being processed.
+    pixel_to_nm_scaling : float
+        Pixel to nm scaling.
+    min_skeleton_size : int
+        Minimum size of grain in pixels after skeletonisation.
+    mask_smoothing_params : dict
+        Dictionary of parameters to smooth the grain mask for better quality skeletonisation results. Contains
+        a gaussian 'sigma' and number of dilation iterations.
+    skeletonisation_params : dict
+        Dictionary of options for skeletonisation, options are 'zhang' (scikit-image) / 'lee' (scikit-image) / 'thin'
+        (scikitimage) or 'topostats' (original TopoStats method).
+    pruning_params : dict
+        Dictionary of options for pruning.
+    pad_width : int
+        Padding to the cropped image mask.
+
+    Returns
+    -------
+    tuple[dict, dict]
```

Comment:
As per above comment four objects are returned and I think they are...

```suggestion
    tuple[dict, pd.DataFrame, dict, pd.DataFrame]
```